### PR TITLE
Build & Deploy with Travis-CI

### PR DESCRIPTION
**Pull request by Blackbaud-SteveBrush**
## Purpose:

To create a workflow for continuous integration deployments with Travis-CI. There should be marginal effort required from our basic--and advanced--Stache users to deploy their sites in a way that is consistent and adheres to best practices. This particular workflow borrows heavily from [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow), and _enforces pull requests as the only method for deployment_.
## Overview:

Azure sites "watch" a Github branch and trigger a deployment when that branch receives a push notification. Currently, Azure will _build and deploy_ the site; this feature branch hopes to remove the build step from Azure to speed up the deployment process and catch any issues that occur during the process. Travis-CI will execute the build and commit the build to the `deploy` branches, which Azure watches.
## Features:
- Added new Stache commands:
  - `stache release` Mirrors `grunt bump` and pushes to origin
  - `stache commitBuild` Commits build results to deploy branch
- Added Travis-CI config and build scripts
### Getting started
1. Ensure that you are the only one reviewing this pull request (some actions below would be destructive to another reviewer).
2. Go to [travis-docs](https://github.com/blackbaud/travis-docs) and delete all branches except `master`.
3. Clone the [travis-docs](https://github.com/blackbaud/travis-docs) repo locally and use it for your tests, since it is already connected to the [Travis-CI dashboard](https://travis-ci.com/blackbaud/travis-docs).
4. Install the Stache CLI `feature-travis-deploy` branch on your local machine, globally (this will replace your existing version of the Stache CLI): 

```
$  npm install -g git+https://github.com/blackbaud/stache-cli.git#feature-travis-deploy
```
## Testing for basic Stache sites:

screen shot 2016-08-26 at 9 47 32 am
1. On the [travis-docs](https://github.com/blackbaud/travis-docs) repo, create a new branch named `deploy` from the `master` branch
2. On the Azure Portal, edit the "blackbaud-travis-docs" Azure App Service, making sure that its deployment source is set to watch the `travis-docs:deploy` branch.
3. Clone [travis-docs](https://github.com/blackbaud/travis-docs) locally and make sure the file **stache.deploy.yml** does NOT exist on the root.
4. Next, create a new branch named `feature`.
5. Create an arbitrary commit and push to the `feature` branch.
6. Create a pull request, `master` <-- `feature`, and merge.
7. Travis-CI should build to the `deploy` folder and make a commit to the `deploy` branch.
8. Azure should trigger a redeploy (but **NOT** a build).
9. Make sure that the [Azure site](http://travis.developer.blackbaud.com/) represents your change.
### Workflow diagram for A-Team/The Flash:

diagram
### Deploying to test site

screen shot 2016-08-26 at 9 47 44 am
- On Github, delete all branches except for `master`.
- Clone [travis-docs](https://github.com/blackbaud/travis-docs) locally and open (or create) **stache.deploy.yml**. The file's contents should mirror:

```
env:
  STACHE_DEPLOY_TEST_BRANCH: deploy-test
  STACHE_DEPLOY_PROD_BRANCH: deploy-prod
  STACHE_MASTER_BRANCH: master
  STACHE_DEVELOP_BRANCH: develop
  STACHE_GITHUB_ORG: blackbaud
  STACHE_BUILD_DIRECTORY: deploy
build: <%= stache.config.env.STACHE_BUILD_DIRECTORY %>/
```
- Commit the changes to this file to the `master` branch.
- After the file has been committed to `master`, create the following branches off of the `master` branch:
  - `deploy-test`
  - `deploy-prod`
  - `develop`
- Edit the "blackbaud-travis-docs" Azure App Service, making sure that its deployment source is set to watch the `travis-docs:deploy-test` branch.
- Next, create a new branch from `develop` named `feature`.
- Create an arbitrary commit and push to the `feature` branch.
- Create a pull request, `develop` <-- `feature`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-test` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure site](http://travis.developer.blackbaud.com/) represents your change.
- Delete the `feature` branch.
### Deploying to production site

screen shot 2016-08-26 at 9 47 50 am
- (This section depends on the "Deploying to test site" above.)
- Edit the "blackbaud-travis-docs" Azure App Service, making sure that its deployment source is set to watch the `travis-docs:deploy-prod` branch (in a real-life scenario, you would have two separate Azure sites for test and production).
- Next, create a new branch from `develop` named `release`.
- Checkout the `release` branch locally and type `stache release`. Stache CLI will create a commit and push up to `origin release`.
- Create a pull request, `master` <-- `release`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-prod` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure production site](http://travis.developer.blackbaud.com/) represents your change.
- Create a pull request, `develop` <-- `release`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-test` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure test site](http://travis.developer.blackbaud.com/) represents your change.
- Delete the `release` branch.
### Hotfixes

screen shot 2016-08-26 at 9 47 57 am
- (This section depends on the "Deploying to production site" above.)
- Create a new branch from `master` named `hotfix`.
- Checkout the `hotfix` branch locally and type `stache release`. Stache CLI will create a commit and push up to `origin hotfix` (in a real-world scenario, an actual change to the source code would come before typing `stache release`).
- Create a pull request, `master` <-- `hotfix`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-prod` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure production site](http://travis.developer.blackbaud.com/) represents your change.
- Create a pull request, `develop` <-- `hotfix`, and merge.
- Travis-CI should build to the `deploy` folder and make a commit to the `deploy-test` branch.
- Azure should trigger a redeploy (but **NOT** a build).
- Make sure that the [Azure test site](http://travis.developer.blackbaud.com/) represents your change.
- Delete the `hotfix` branch.

### CODE REVIEW

**Comment by Blackbaud-SteveBrush** [comment]
@blackbaud-brandonhare: I'm going to let you take a look at this first, before we get other teams involved. Let me know if I missed anything (probably did), or if you need to get on the phone to walk through it; it's a doozy. Remember: both standard Stache users and large-scale engineering teams need to benefit from this model.

**Comment by blackbaud-brandonhare** [comment]
So when running the deploy to production tests, the deploy-test content gets updated, but the Azure redeploy won't get triggered because it isn't watching that branch.

**Comment by blackbaud-brandonhare** [comment]
Same for hotfix.

File: stache-boilerplate/.travis.yml-sample [hunk 1]
```diff
@@ -0,0 +1,12 @@
+language: node_js
+
+before_script:
+  - npm i -g grunt-cli blackbaud-stache-cli
+
+script:
+  - stache build --config=stache.yml,stache.prod.yml,stache.deploy.yml
```
Comment (blackbaud-brandonhare) [inline stache-boilerplate/.travis.yml-sample]: With this script update, will we still be able to build locally? Or is the only build process now through github/Travis? All of the User Ed how-to pages are currently manually deployed to the bbweb server. These are the only projects I know of that deploy this way though.

File: scripts/commit-build.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# No need to commit the build during a pull request.
+if [[ "$TRAVIS_EVENT_TYPE" == "pull_request" ]]; then
+  exit 0
+fi
+
+# Capture the last commit's message
+LAST_COMMIT_MESSAGE=`git log --format=%B -n 1 $TRAVIS_COMMIT`
+
+# Is the commit asking to be deployed?
+REGEX_DEPLOY_COMMENT="(^Merge pull request).*(${STACHE_GITHUB_ORG}\/release|${STACHE_GITHUB_ORG}\/hotfix)"
```
Comment (blackbaud-brandonhare) [inline scripts/commit-build.sh:42]: Regex with variables?! _SWOON_

File: scripts/commit-build.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  commit_build $STACHE_DEPLOY_TEST_BRANCH
+
+# Is the base branch the master branch?
+elif [[ "$TRAVIS_BRANCH" == "$STACHE_MASTER_BRANCH" ]]; then
+
+  # Is this commit a release or hotfix?
+  if [[ $LAST_COMMIT_MESSAGE =~ $REGEX_DEPLOY_COMMENT ]]; then
+    commit_build $STACHE_DEPLOY_PROD_BRANCH
+  fi
+fi
```
Comment (blackbaud-brandonhare) [inline scripts/commit-build.sh:88]: This is pretty cool. Bash scripting syntax is weird though.

File: stache-boilerplate/.travis.yml-sample (hunk 1 above)
Comment (Blackbaud-SteveBrush) [reply stache-boilerplate/.travis.yml-sample]: Building locally should be unaffected. Existing, internal sites should also be unaffected (would be worth testing this).

File: scripts/commit-build.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+git config --global user.email "stache-build-user@blackbaud.com"
+git config --global user.name "Blackbaud Stache Build User"
+
+commit_build() {
+  echo "Committing build results to ${1}...";
+  git status
+  git add --all
+  git stash save
+  git checkout $1 --quiet || git checkout -b $1
+  git stash pop
```
Comment (blackbaud-brandonhare) [inline scripts/commit-build.sh]: Just curious. Why are we doing stashing here? Why not a merge?

**Comment by blackbaud-brandonhare** [comment]
I think this is killer Steve. I didn't really see anything that stood out to me as problematic. Well structured, well organized, and well coded. As usual. :) Well done!

**Comment by Blackbaud-SteveBrush** [comment]
@Blackbaud-BobbyEarl: Thanks for your feedback regarding the default environment variables. Since most of the `env` variables in the **stache.deploy.yml** file were being used by the `scripts/commit-build.sh` file, I added the default checks there. I also removed the requirement for **stache.deploy.yml**; the update only requires the **.travis.yml** file. Let me know if this is acceptable!
(Would also like your feedback, @blackbaud-brandonhare!)

**Comment by blackbaud-brandonhare** [comment]
This version is better than the last one. I didn't really notice anything that stood out. Great stuff!

**Pull request merged**